### ISSUE TITLE
Compiler should not warn about operator precedence by default

### ISSUE THREAD

**Issue by d6fd0808-f213-4703-8fea-222e641ecd67** [llvm/llvm-project#30763]
|  |  |
| --- | --- |
| Bugzilla Link | [31415](https://llvm.org/bz31415) |
| Version | 3.9 |
| OS | All |
| CC | @hfinkel |

## Extended Description 
clang has a number of diagnostics warning about code that uses the C operator precedence table in a way some programmers misunderstand. This category of warnings is occasionally useful in catching bugs.

However, these warnings should not be enabled by default (i.e. without any -W... switched) as they usually do not indicate incorrect code. Indeed, full usage of the C operator precedence rules is *not* indicative of a wrong program and clang warning about that causes an exceptional amount of noise during compilation of perfectly fine code. As by the guideline that all default warnings should only be warnings that indicate serious problems or likely mistakes, I believe that operator precedence warnings should not be enabled by default. I request that they be disabled by default and only enabled if explicitly requested (e.g. with -Wall).

**Comment by Quuxplusone** [comment]
This issue is basically "Some warning has false positives," but without showing the code that triggers the false positive or even naming the `-Wfoo` option involved. Without any specific examples, I think it's impossible to act on this issue.

**Comment by clausecker** [comment]
Thanks for closing this issue immediately instead of, you know, asking the author to elaborate with details.

**Comment by clausecker** [comment]
The specific warning in this case is `-Wparentheses`, a default warning.  The warning is apparently further split up into various additional warnings, but `-Wno-parentheses` turns them all off.  It produces bogus warnings like this:

```
./tileset.h:190:69: warning: '&' within '|' [-Wbitwise-op-parentheses]
        return (eq & (c | c  << 5 | (c & 0x0f7bdef) << 1 | c >> 5 | c >> 1 & 0x0f7bdef));
                                                                  ~ ~~~~~~~^~~~~~~~~~~
```

for perfectly fine code like this:

```
typedef unsigned int tileset;
static inline tileset
tileset_reduce_eqclass(tileset eq)
{
	tileset c = tileset_complement(eq);

	/*
	 * the mask prevents carry into other rows:
	 * 0x0f7bdef: 01111 01111 01111 01111 01111
	 */
	return (eq & (c | c  << 5 | (c & 0x0f7bdef) << 1 | c >> 5 | c >> 1 & 0x0f7bdef));
}
```

From my project [24puzzle](https://github.com/clausecker/24puzzle).

The compiler should not produce warnings for correct code like this.

@Quuxplusone Please let me know if this is the kind of example you were looking for and reopen the issue if it is.

**Comment by dwblaikie** [comment]
> The specific warning in this case is -Wparentheses, a default warning.
(for what it's worth, `-Wbitwise-op-parentheses` is not enabled by default so far as I can tell, if that's what you mean by "a default warning" - you'd have to pass some warning flag (`-Wall`, `-Wextra`, `-Wparentheses`, `-Wbitwise-op-parentheses`, etc) to be getting that warning)

So I think the request "I request that they be disabled by default and only enabled if explicitly requested (e.g. with -Wall)." is already addressed/that this is the current/existing behavior.

https://godbolt.org/z/sfEYaj9rq

**Comment by Quuxplusone** [comment]
Confirmed: neither Clang nor GCC warn for that snippet.
It looks like Clang 9.x did warn by default, but that was changed in Clang 10, and Clang no longer warns on that snippet (unless you pass `-Wall`). https://godbolt.org/z/Tonnqx5c7
